synapse/lib/snap.py

```diff
@@ -1388,6 +1393,129 @@ async def addNode(self, name, valu, props=None, norminfo=None):
         # the newly constructed node is cached
         return await self.getNodeByBuid(protonode.buid)
 
+    async def _addGuidNodeByDict(self, form, vals, props=None):
+
+        norms = {}
+        counts = []
+        proplist = []
+
+        if props is None:
+            props = {}
+
+        trycast = vals.pop('$try', False)
+        addprops = vals.pop('$props', None)
+        if addprops is not None:
+            props.update(addprops)
+
+        try:
+            for name, valu in list(props.items()):
+                props[name] = form.reqProp(name).type.norm(valu)
+
+            for name, valu in vals.items():
+
+                prop = form.reqProp(name)
+                norm, norminfo = prop.type.norm(valu)
+
+                norms[name] = (prop, norm, norminfo)
+                proplist.append((name, norm))
+        except s_exc.BadTypeValu as e:
+            if not trycast: raise
+            await self.warn(f'Bad type value: {e}')
```

Comment:
Should we include the name of the prop here as well? Using `?=` should be catching `BadTypeValu` in the AST so I think the additional warning would be the only benefit of using `$try` instead.